airbyte-lib/airbyte_lib/executor.py

```diff
@@ -0,0 +1,155 @@
+            raise Exception(f"Process exited with code {exit_code}")
+
+
+class VenvExecutor(Executor):
+    def __init__(self, metadata: ConnectorMetadata, target_version: str = "latest", install_if_missing: bool = False):
+        super().__init__(metadata, target_version)
+        self.install_if_missing = install_if_missing
+
+    def _get_venv_name(self):
+        return f".venv-{self.metadata.name}"
+
+    def _get_connector_path(self):
+        return Path(self._get_venv_name(), "bin", self.metadata.name)
+
+    def _run_subprocess_and_raise_on_failure(self, args: List[str]):
+        result = subprocess.run(args)
+        if result.returncode != 0:
+            raise Exception(f"Install process exited with code {result.returncode}")
+
+    def install(self):
+        venv_name = self._get_venv_name()
+        self._run_subprocess_and_raise_on_failure([sys.executable, "-m", "venv", venv_name])
+
+        pip_path = os.path.join(venv_name, "bin", "pip")
+
+        # TODO this is a temporary install path that will be replaced with a proper package name once they are published. At this point we are also using the version
+        package_to_install = f"../airbyte-integrations/connectors/{self.metadata.name}"
+        self._run_subprocess_and_raise_on_failure([pip_path, "install", "-e", package_to_install])
+
+    def _get_installed_version(self):
+        """
+        In the venv, run the following: python -c "from importlib.metadata import version; print(version('<connector-name>'))"
+        """
+        venv_name = self._get_venv_name()
+        connector_name = self.metadata.name
+        return subprocess.check_output(
+            [os.path.join(venv_name, "bin", "python"), "-c", f"from importlib.metadata import version; print(version('{connector_name}'))"],
+            universal_newlines=True,
+        ).strip()
+
+    def ensure_installation(self):
+        venv_name = f".venv-{self.metadata.name}"
+        venv_path = Path(venv_name)
+        if not venv_path.exists():
+            if not self.install_if_missing:
+                raise Exception(f"Connector {self.metadata.name} is not available - venv {venv_name} does not exist")
+            self.install()
+
+        connector_path = self._get_connector_path()
+        if not connector_path.exists():
+            raise Exception(f"Could not find connector {self.metadata.name} in venv {venv_name}")
+
+        installed_version = self._get_installed_version()
+        if installed_version != self.target_version:
+            # If the version doesn't match, reinstall
+            self.install()
+
+            # Check the version again
+            version_after_install = self._get_installed_version()
+            if version_after_install != self.target_version:
+                raise Exception(
+                    f"Failed to install connector {self.metadata.name} version {self.target_version}. Installed version is {version_after_install}"
+                )
```

Comment:
In a follow-on PR, I suggest a flag to make version verification an opt-in behavior, where version would not be checked by default, but could be checked/auto-fixed as needed.

- https://github.com/airbytehq/airbyte/pull/33878/files#diff-719b4672d504b4121f32fe105e9d2dead848338b3e094f00dd316f6452552487R169-R180

Reply:
It will get less relevant once we publish properly to pypi, but I can see how the current state might be too restrictive or take too much control from the user.

What some tools are doing is to log a warning if the version doesn't match, but don't fail right away. Seems suitable here, but nothing we can't adjust before the first release.

Reply:
> What some tools are doing is to log a warning if the version doesn't match, but don't fail right away. Seems suitable here, but nothing we can't adjust before the first release.

Totally! Yeah, I think a warning is a nice balance - gives the user information but doesn't necessary cause direct failure. In that case, we could always check version, and the "opt in" part would be whether to hard-fail, rather than warn. I can imagine a hard-failure might be preferred in specific instances where we think we ware explicitly requesting a particular version number.